HSEARCH-3107 Test performance with a complex entity model

fax4ever opened the pull request:
https://hibernate.atlassian.net/browse/HSEARCH-3107

yrodiere commented on search6/src/main/java/org/hibernate/performance/search/model/application/SearchProgrammaticMapping.java:
```diff
@@ -0,0 +1,128 @@
+		question.property( "text" ).fullTextField().analyzer( "default" );
+
+		TypeMappingStep openQuestion = mapping.type( OpenQuestion.class );
+		openQuestion.indexed();
```
`indexed()` is inherited in Search 6, so this is not necessary.

yrodiere commented on search6/src/main/java/org/hibernate/performance/search/model/application/SearchProgrammaticMapping.java:
```diff
@@ -0,0 +1,128 @@
+		openQuestion.indexed();
+
+		TypeMappingStep closedQuestion = mapping.type( ClosedQuestion.class );
+		closedQuestion.indexed();
```
`indexed()` is inherited in Search 6, so this is not necessary.

yrodiere commented on search5/src/main/java/org/hibernate/performance/search/model/application/SearchProgrammaticMapping.java:
```diff
@@ -0,0 +1,149 @@
+		IndexedMapping performanceSummary = mapping.entity( PerformanceSummary.class ).indexed();
+		performanceSummary
+				.property( "id", ElementType.FIELD ).documentId()
+				.property( "employee", ElementType.FIELD ).indexEmbedded().depth( 3 )
```
If you do this, I think there should be a `containedIn()` from `Employee` to `PerformanceSummary`. Employees are expected to have their data change from time to time, and it wouldn't cost much to reindex their perf summaries when that happens.

yrodiere replied:
Also, maybe you could use `includePaths` here? You probably don't need to index-embed *everything*, and I know for a fact that when using `includePaths`, Search 6 will be smarter and skip reindexing in some cases.

yrodiere commented on search5/src/main/java/org/hibernate/performance/search/model/application/SearchProgrammaticMapping.java:
```diff
@@ -0,0 +1,149 @@
+				.property( "uniqueCode", ElementType.FIELD ).field().analyze( Analyze.NO )
+				.property( "definition", ElementType.FIELD ).indexEmbedded()
+				.property( "approval", ElementType.FIELD ).indexEmbedded()
+				.property( "subject", ElementType.FIELD ).indexEmbedded()
```
If you do this, I think there should be a `containedIn()` from `Employee` to `QuestionnaireInstance`. Employees are expected to have their data change from time to time, and it wouldn't cost much to reindex their questionnaires when that happens.

yrodiere commented on model/src/main/java/org/hibernate/performance/search/model/entity/Employee.java:
```diff
@@ -0,0 +1,92 @@
+@Entity
+public class Employee extends IdEntity {
+
+	protected String name;
```
Is this the first name? If so, `firstName` seems clearer.

yrodiere commented on search6/src/main/java/org/hibernate/performance/search/model/application/SearchProgrammaticMapping.java:
```diff
@@ -0,0 +1,128 @@
+		TypeMappingStep performanceSummary = mapping.type( PerformanceSummary.class );
+		performanceSummary.indexed();
+		performanceSummary.property( "id" ).genericField();
+		performanceSummary.property( "employee" ).indexedEmbedded().includeDepth( 3 )
```
Remember to add `includePaths` here if you do it in Search 5.

yrodiere commented on search6/src/main/java/org/hibernate/performance/search/model/application/SearchProgrammaticMapping.java:
```diff
@@ -0,0 +1,128 @@
+		questionnaireInstance.property( "approval" ).indexedEmbedded().indexingDependency()
+				.reindexOnUpdate( ReindexOnUpdate.NO );
+		questionnaireInstance.property( "subject" ).indexedEmbedded().indexingDependency()
+				.reindexOnUpdate( ReindexOnUpdate.NO );
```
Remember to remove this if you add a `containedIn()` in Search 5.

yrodiere commented on search6/src/main/java/org/hibernate/performance/search/model/application/SearchProgrammaticMapping.java:
```diff
@@ -0,0 +1,128 @@
+		performanceSummary.indexed();
+		performanceSummary.property( "id" ).genericField();
+		performanceSummary.property( "employee" ).indexedEmbedded().includeDepth( 3 )
+				.indexingDependency().reindexOnUpdate( ReindexOnUpdate.NO );
```
Remember to remove this if you add a `containedIn()` in Search 5.

yrodiere commented on search6/src/main/java/org/hibernate/performance/search/model/application/ModelServiceImpl.java:
```diff
@@ -0,0 +1,127 @@
+				AutomaticIndexingSynchronizationStrategyNames.WRITE_SYNC );
+
+		if ( kind.isLucene() ) {
+			config.put( "hibernate.search.backend.directory.type", "local-heap" );
```
Same as Search 5: this should be `filesystem`, and you should provide a root path for the indexes.

yrodiere commented on search6/src/main/java/org/hibernate/performance/search/model/application/ModelServiceImpl.java:
```diff
@@ -0,0 +1,127 @@
+	@Override
+	public <E> List<E> searchAnd(Session session, Class<E> entityClass, String fieldName1, Object value1,
+			String fieldName2, Object value2) {
+		SearchSession searchSession = Search.session( session );
+
+		SearchScope<E> scope = searchSession.scope( entityClass );
+		MatchPredicateOptionsStep<?> matching1 = scope.predicate().match().field( fieldName1 ).matching( value1 );
+		MatchPredicateOptionsStep<?> matching2 = scope.predicate().match().field( fieldName2 ).matching( value2 );
+		SearchPredicate predicate = scope.predicate().bool().must( matching1 ).must( matching2 ).toPredicate();
+
+		return searchSession.search( scope ).where( predicate ).fetchHits( DEFAULT_LIMIT );
```
This would be easier on the eye:

```suggestion
		return Search.session( session ).search( entityClass )
				.where( f -> f.bool()
						.must( f.match().field( fieldName1 ).matching( value1 ) )
						.must( f.match().field( fieldName2 ).matching( value2 ) ) )
				.fetchHits( DEFAULT_LIMIT );
```

yrodiere commented on search6/src/main/java/org/hibernate/performance/search/model/application/ModelServiceImpl.java:
```diff
@@ -0,0 +1,127 @@
+		return Search.session( session ).search( entityClass ).where(
+				f -> f.range().field( fieldName )
+						// include limits
+						.between( start, RangeBoundInclusion.INCLUDED, end, RangeBoundInclusion.INCLUDED ) )
```
Or simply:

```suggestion
						.between( start, end ) )
```

yrodiere commented on search6/src/main/java/org/hibernate/performance/search/model/application/ModelServiceImpl.java:
```diff
@@ -0,0 +1,127 @@
+		return Search.session( session ).search( entityClass ).where(
+				f -> f.range().field( fieldName )
+						// include limits
+						.between( start, RangeBoundInclusion.INCLUDED, end, RangeBoundInclusion.INCLUDED ) )
```
```suggestion
						.between( start, end ) )
```

yrodiere commented on search6/src/main/java/org/hibernate/performance/search/model/application/ModelServiceImpl.java:
```diff
@@ -0,0 +1,127 @@
+
+	@Override
+	public List<Object> projectId(Session session, Class<?> entityClass, String fieldName, Object value) {
+		List<EntityReference> entityReferences = Search.session( session ).search( entityClass )
+				.selectEntityReference()
+				.where( f -> f.match().field( fieldName ).matching( value ) )
+				.fetchHits( DEFAULT_LIMIT );
+
+		return entityReferences.stream().map( a -> a.id() ).collect( Collectors.toList() );
```
To avoid creating multiple lists:

```suggestion
		return Search.session( session ).search( entityClass )
				.select( f -> f.composite( ref -> ref.id(), f.entityReference() ) )
				.where( f -> f.match().field( fieldName ).matching( value ) )
				.fetchHits( DEFAULT_LIMIT );
```

yrodiere commented on search6/src/main/java/org/hibernate/performance/search/model/application/ModelServiceImpl.java:
```diff
@@ -0,0 +1,127 @@
+	@Override
+	public List<List<?>> project(Session session, Class<?> entityClass, String fieldName1, Object value1,
+			String fieldName2, Object value2, String projectedField1, String projectedField2) {
+		SearchSession searchSession = Search.session( session );
+
+		SearchScope<?> scope = searchSession.scope( entityClass );
+
+		FieldProjectionValueStep<?, Object> projection1 = scope.projection().field( projectedField1 );
+		FieldProjectionValueStep<?, Object> projection2 = scope.projection().field( projectedField2 );
+		SearchProjection<List<?>> projection = scope.projection().composite( projection1, projection2 )
+				.toProjection();
+
+		MatchPredicateOptionsStep<?> matching1 = scope.predicate().match().field( fieldName1 ).matching( value1 );
+		MatchPredicateOptionsStep<?> matching2 = scope.predicate().match().field( fieldName2 ).matching( value2 );
+		SearchPredicate predicate = scope.predicate().bool().must( matching1 ).must( matching2 ).toPredicate();
+
+		return searchSession.search( scope ).select( projection ).where( predicate ).fetchHits( DEFAULT_LIMIT );
```
Or simply:

```suggestion

		return Search.session( session ).search( entityClass )
				.select( f -> f.composite( f.field( projectedField1 ), f.field( projectedField2 ) ) )
				.where( f -> f.bool()
						.must( f.match().field( fieldName1 ).matching( value1 ) )
						.must( f.match().field( fieldName2 ).matching( value2 ) ) )
				.fetchHits( DEFAULT_LIMIT );
```

yrodiere reviewed (commented):
Here's some feedback. Overall this is a good start, but for the automatic indexing benchmarks, I think we need to focus on "real-world", everyday operations, especially for inserts. And as I explained below, we definitely need to make sure that every invocation will do the same thing, especially for deletes, otherwise the measurements are just wrong.

fax4ever commented on jenkins/performance-search5-elasticsearch-scripted.groovy:
```diff
@@ -0,0 +1,154 @@
+ * License: GNU Lesser General Public License (LGPL), version 2.1 or later
+ * See the lgpl.txt file in the root directory or <http://www.gnu.org/licenses/lgpl-2.1.html>.
+ */
+
```
I created both to assess which one was the better one. I would keep the declarative one if it would be possible.

fax4ever commented on jenkins/performance-search5-elasticsearch.groovy:
```diff
@@ -0,0 +1,61 @@
+                    sh """ \
+					java \
+					-jar benchmarks.jar \
+					-jvmArgsAppend -Dhosts=vpc-ci-elasticsearch-52-r5-membawptnueqlpmzkr677lwzoa.us-east-1.es.amazonaws.com \
```
Sure! There was just a test...

fax4ever commented on jenkins/performance-search6-elasticsearch-scripted.groovy:
```diff
@@ -0,0 +1,154 @@
+ * License: GNU Lesser General Public License (LGPL), version 2.1 or later
+ * See the lgpl.txt file in the root directory or <http://www.gnu.org/licenses/lgpl-2.1.html>.
+ */
+
```
only one will remain

fax4ever commented:
Thank you @yrodiere for the extensive and very good review.
So far I've not found any comment with which I didn't agree with.
I pushed some commits, but the fix is not finished, I left 9 comments without any feed at the moment.
I let you know when I finish it.

fax4ever commented on model/src/main/java/org/hibernate/performance/search/model/application/DomainDataFiller.java:
```diff
@@ -0,0 +1,85 @@
+		} );
+
+		// Phase 5: simulate the employees filling the questionnaires
+		new AnswerFiller( sessionFactory ).fillAllAnswers();
```
Sure!

fax4ever commented on model/src/main/java/org/hibernate/performance/search/model/asset/AutomaticIndexingState.java:
```diff
@@ -0,0 +1,148 @@
+
+	public AutomaticIndexingUpdatePartitionState getUpdatePartition(int threadNumber) {
+		checkThreadNumber( threadNumber );
+		synchronized (this) {
```
The idea was to use the data created by `test1_insert` in the subsequent tests (updates and deletes). This makes the creation of the `indexUpdatePartitions` and `indexDeletePartitions` dependent by the number of iterations/invocations performed by `test1_insert`.
I agree. That was not a good idea.
Let's make `indexUpdatePartitions` and `indexDeletePartitions` using only the data created by `initialIndexSize`.

fax4ever replied:
This fix will allow us to create `indexUpdatePartitions` and `indexDeletePartitions` eagerly.

fax4ever commented on model/src/main/java/org/hibernate/performance/search/model/application/DomainDataFiller.java:
```diff
@@ -0,0 +1,85 @@
+
+	public void fillData(int companyId) {
+		// atomicity is not used here
+		AtomicReference<Company> companyReference = new AtomicReference<>();
```
+1 about the find

about having a single transaction:
if the relationship size is `LARGE` we have more that 100_000 inserts for each invocation of this method, from here the idea of introducing multiple transactions.

fax4ever commented:
@yrodiere thank you again for the extensive review.
I should have addressed all the comments here.

fax4ever commented:
I've fixed some stuff: on system properties, session factory closing and a wrong method invocation.
It seems now that the performance tests are run correctly. E.g. see https://ci.hibernate.org/job/hibernate-search-performance/job/search6-elasticsearch/21/console.
So I'm finally ready for your new review. @yrodiere

yrodiere commented on jmh-tck/src/main/java/org/hibernate/performance/search/tck/AutomaticIndexingStateHolder.java:
```diff
@@ -0,0 +1,51 @@
+	private RelationshipSize relationshipSize;
+
+	@Param({ "100" })
+	private int initialIndexSize;
```
I think this should be `initialCompanyCount`, like in `MassIndexingPerformanceTest` and `SearchingPerformanceTest`?

yrodiere commented on model/src/main/java/org/hibernate/performance/search/model/asset/AutomaticIndexingState.java:
```diff
@@ -43,98 +46,95 @@ public AutomaticIndexingState(RelationshipSize relationshipSize, int initialInde
+		if ( !started ) {
+			return;
+		}
+		new DomainDataRemover( sessionFactory ).truncateAll();
```
If you're going to do this, you also need to initialize the data when the iteration starts.
Otherwise, the first (warmup) iteration will work on a large database and index, and the next ones will work on an almost-empty database and index.

Another approach would be to keep the data between iterations, but I don't know how feasible that would be. It's probably too complex, so I wouldn't try unless initializing the data is very slow.

fax4ever replied:
I think that we already do that if we start an iteration : iteration number > 1. Is that possible?

yrodiere replied:
You're right. Somehow I didn't notice that you call `stop` after each iteration, leading to `start` being executed before each iteration, leading to data initialization before each iteration.

yrodiere reviewed (commented):
Thanks! Here is another round of comments.

This time, my main concern is the score error in the CSV reports. As you can see, some scores have 99.9% confidence intervals ranging from 0 to 2 times the score, which is huge. There must be something wrong somewhere. I added a few comments where I found something suspicious, either technically (lack of connection pool?) or on the "business" side (benchmarking an operation that changes the manager of an employee AND changes its name AND changes the name of the manager: not very representative of real-world operations).

I suggest you address these problems, and then move the repo to the Hibernate organization: we'll iterate from there. If the score errors are still high, we'll need to address that first, otherwise the performance tests are not reliable enough to be of use.

By the way, one possible reason for the score errors to be high might be simply that the benchmarks didn't run for long enough. If that's the case, we will need to consider running each benchmark longer, and cutting down the number of benchmarks to avoid needing 2+ hours for each run.

yrodiere commented on jenkins/performance-search5-lucene.groovy:
```diff
@@ -0,0 +1,52 @@
+pipeline {
+    agent any
```
We have separate nodes for performance tests, please use those.

```suggestion
    agent 'Performance'
```

fax4ever replied:
If I try to apply the change, I got the following error: https://ci.hibernate.org/job/hibernate-search-performance/job/search5-lucene/10/console.

I'm finding another way to chose the node.

fax4ever replied:
The following seems to be working:

``` groovy
agent {
  label 'Performance'
}
```

fax4ever commented on model/src/main/java/org/hibernate/performance/search/model/asset/AutomaticIndexingState.java:
```diff
@@ -43,98 +46,95 @@ public AutomaticIndexingState(RelationshipSize relationshipSize, int initialInde
+		if ( started ) {
+			return;
+		}
+		start();
```
I think that it is executed if we start iteration : iteration number > 1. Do you agree?

yrodiere replied:
Ah, I see, you call `stop` after each iteration. `start` doesn't start Hibernate Search, it generates partitions and such.

In that case, I'm not entirely sure why you call `start` in `startTrial()` and `stop` in `stopTrial`?

fax4ever replied:
You're right we can call it only from `startIteration` and `stopIteration`, since it seems that `startIteration` is always called after `startTrial` and `stopIteration` is always called just before `stopTrial`.

fax4ever commented on model/src/main/java/org/hibernate/performance/search/model/asset/AutomaticIndexingDeleteInsertPartitionState.java:
```diff
@@ -0,0 +1,125 @@
+		this.initialIndexSize = initialIndexSize;
+		this.numberOfThreads = numberOfThreads;
+		this.threadNumber = threadNumber;
+		this.invocationSize = invocationSize;
```
Going to remove `invocationSize`... Basically it will work as it was always 1.

fax4ever commented on model/src/main/java/org/hibernate/performance/search/model/asset/AutomaticIndexingUpdatePartitionState.java:
```diff
@@ -36,27 +39,35 @@ public AutomaticIndexingUpdatePartitionState(SessionFactory sessionFactory,
-			updateCompanyBUOneTime();
-		}
+		domainDataUpdater.inTransaction( ( (session, up) -> {
+			for ( int i = 0; i < invocationSize; i++ ) {
```
Going to remove `invocationSize`... Basically it will work as it was always 1.

yrodiere commented on jenkins/performance-search5-elasticsearch.groovy:
```diff
@@ -1,136 +1,73 @@
 							-jvmArgsAppend -Dhibernate.search.default.elasticsearch.aws.access_key=$AWS_ACCESS_KEY_ID \
 							-jvmArgsAppend -Dhibernate.search.default.elasticsearch.aws.secret_key=$AWS_SECRET_ACCESS_KEY \
-							-wi 1 -i 3 \
+							-wi 0 -i 2 \
```
Why is this down to 0 warmup iterations and 2 measurement iterations?

If we want less score errors, we need:

* some warmup
* as many iterations as we can
* iterations that are as long as possible

Please try to adjust these settings to get a better score error.

fax4ever replied:
I added this commit only to make the execution of the performance tests shorter. Let's start by removing it and arrange them in the Hibernate organization repository.

yrodiere commented on model/src/main/java/org/hibernate/performance/search/model/application/DomainDataUpdater.java:
```diff
@@ -139,16 +137,29 @@ public void updateQuestionsAndAnswers(Session session, int invocation, int quest
 		List<QuestionnaireInstance> instances = new EmployeeRepository( session ).findByDefinition( definition );
-		for ( QuestionnaireInstance instance : instances ) {
-			for ( ClosedAnswer answer : instance.getClosedAnswers() ) {
+		for ( QuestionnaireInstance instance : choose( instances, 2 ) ) {
```
Wouldn't it be more reasonable to include this filtering in the `findByDefinition` query above? From what I understand you may very well be loading hundreds of Questionnaire instances, only to use 2.

fax4ever replied:
I agree, I'm changing it...

yrodiere commented on model/src/main/java/org/hibernate/performance/search/model/application/DomainDataUpdater.java:
```diff
@@ -120,7 +118,7 @@ public void updateQuestionnaire(Session session, int invocation, int questionnai
 
 		List<QuestionnaireInstance> instances = new EmployeeRepository( session ).findByDefinition( definition );
-		for ( QuestionnaireInstance instance : instances ) {
+		for ( QuestionnaireInstance instance : choose( instances, 3 ) ) {
```
Wouldn't it be more reasonable to include this filtering in the `findByDefinition` query above? From what I understand you may very well be loading hundreds of Questionnaire instances, only to use 2.

fax4ever replied:
I agree, I'm changing it...

yrodiere commented on model/src/main/java/org/hibernate/performance/search/model/application/DomainDataUpdater.java:
```diff
@@ -139,16 +137,29 @@ public void updateQuestionsAndAnswers(Session session, int invocation, int quest
-		for ( QuestionnaireInstance instance : instances ) {
-			for ( ClosedAnswer answer : instance.getClosedAnswers() ) {
+		for ( QuestionnaireInstance instance : choose( instances, 2 ) ) {
+			for ( ClosedAnswer answer : choose( instance.getClosedAnswers(), 1 ) ) {
```
Here's it's fine, since I assume there aren't hundreds of answers in a single questionnaire instance.

yrodiere reviewed (commented):
Ok, so the `NaN` scores, that's something new...

Anyway I added a few comments, but let's create the Hibernate org repository, and address score issues after that, if you agree?

fax4ever commented:
> Ok, so the `NaN` scores, that's something new...
> 
> Anyway I added a few comments, but let's create the Hibernate org repository, and address score issues after that, if you agree?

I am definitely agree. I'm addressing the comments...

fax4ever commented:
I think I've done. Let me know if I can create a Hibernate org repository from this branch.

yrodiere commented:
Sure. Just push everything to the main branch, then go to settings and transfer ownership to the Hibernate org.

fax4ever commented:
Merged thanks! Going to change the ownership of this repository.
